### ISSUE TITLE
Create "Page Borders" Augmentation

### ISSUE THREAD

**Issue by jboarman** [sparkfish/augraphy#13]
Pages that are scanned, copied or faxed often are not scanned in perfect alignment or with their borders properly cropped.  This augmentation would attempt to reproduce those poorly scanned edges.

![image](https://user-images.githubusercontent.com/1108065/124686708-30802880-de99-11eb-8ef1-ab8adfb04611.png)
![image](https://user-images.githubusercontent.com/1108065/124686423-bb145800-de98-11eb-8402-7d54f8187185.png)
![image](https://user-images.githubusercontent.com/1108065/124686534-ef881400-de98-11eb-9cf2-a567f3611c52.png)
![image](https://user-images.githubusercontent.com/1108065/124686550-f6af2200-de98-11eb-8a44-c8d5789cbe55.png)
![image](https://user-images.githubusercontent.com/1108065/124686733-3d048100-de99-11eb-9f42-21044c6e1ba9.png)

**Comment by kwcckw** [comment]
I think this could be done by adding some small angle rotation, and heavy dusty ink augmentation at the edges of the image. This is the initial trial without edges based augmentation:

![image](https://user-images.githubusercontent.com/52521979/124689204-abcdf280-df0a-11eb-828f-bd16e4145d10.png)

To improve the visualization , adding edges based dusty ink and white noise should be helpful. I will post again with updated results later.

**Comment by jboarman** [comment]
* **Random Noise** -- I agree that there is a certain level of random noise that appears around the edges.  That noise tends to be associated depth caused by additional pages being included in the scan.  That noise needs to be limited to the edge work (not the center of the page) and should vary since it's not always present..
* **Background** -- So far, in the examples that I was able to locate, the background does not seem to have much noise and appears to universally be black within these samples.  In my own experience, however, I'm certain that I've seen the background as white, with only the edge of the paper being visible as somewhat straight lines representing the edges of the page.  So, I'm thinking that the background could be randomly black or what, so long as the page edge is always visible.  I think the times when it is black is again caused by depth, such as copying from a book on a copy machine, which causes the copy machine cover to be raised and out of focus for the scanner.
* **Page Edges** -- As noted above, for pages on a white background, we'll need semi-straight lines that represent the edges of the page.  So, we'll need to start with a thin border created around the perfectly straight lines (created from the edge of the image prior to rotation) and slightly degrade those edges in a natural-looking manner and possibly with semi-rounded corners.
* **Extra Pages** -- Often, there are additional lines or pages represented which are similarly rotated and offset from the main page.
* **Page Curl Effect** -- I think we may need to address page curl in a separate issue as it may be too complicated to address in this issue.  The effect is that whatever content that may be present at the edges (such as the side of the page that bends inward towards the book's binding) gets out of focus and may accumulate additional noise.  That noise can turn into solid black after a threshold of noise in that bend is reached.
* **Black & White** -- It is nice that color can be retained from source images and from template page background images.  However, if we are trying to replicate common copy machines, faxes and office-oriented scanning, then we will need to have a functional pipeline or augmentation that reproduces that effect, regardless of color being present in source images.  So, perhaps we need a separate Augmentation that does B/W and Greyscaling of images (which seems silly on it's own b/c it's perhaps too simple).  I raise that here to invite conversation for how we might create a new feature request around this need (augmentation vs predefined functional pipeline).  For the purpose of this issue, we should experiment with only B/W images to make it easier to compare to sample images.

**Comment by shaheryar1** [comment]
@jboarman  you can assign this card to me , I am pretty confident that I can implementation this augmentation (except for Page Curl Effect part).

**Comment by jboarman** [comment]
I agree ... we should leave the Page Curl effect to another issue to implement later, probably after #21 (page fold effect) since we can take some parts of that to apply to the page curl logic.

**Comment by shaheryar1** [comment]
There are some results that I got from initial experiments. I am planning to **add ink bleed effect** in the perpendicular direction to lines and also some **random noise around the edges** of semi-straight lines. Obviously it can be followed by random rotation too.

The random parameters I am thinking about
`border_side = [top,left,right,bottom]` - can select multiple randomly
`diminishing_effect=True/False` - will cause the lines to be diminished at the end of page
`pages=[2,3,4,5,6]` - randomly selected. represents the width of the border 

Suggestions are appreciated.


![download (4)](https://user-images.githubusercontent.com/43179913/125596344-ee7810b7-edfe-4983-a95e-c82adf5f367b.png)
![download (5)](https://user-images.githubusercontent.com/43179913/125597112-bdef3454-ce20-444c-b03f-b851e4953b35.png)
![download](https://user-images.githubusercontent.com/43179913/125597170-7119de33-93a9-4ef5-94f7-e08803232f20.png)

**Comment by kwcckw** [comment]
i think the effect looks good , and could you show an example with more than 1 border side?

**Comment by shaheryar1** [comment]
Sure here it is. I also smoothened out the border which makes it more realistic. Also make sure you view the images with github "light theme" to clearly see the borders

![download (7)](https://user-images.githubusercontent.com/43179913/125765555-87ae4972-7fb3-475d-9108-ad9307741d71.png)
![download (9)](https://user-images.githubusercontent.com/43179913/125765481-5884ecd9-fa10-43e6-8496-7a1f2884a536.png)

**Comment by shaheryar1** [comment]
Here are some examples with some noise and offset added to the border

![download (10)](https://user-images.githubusercontent.com/43179913/125775495-f61b729e-0127-43cd-8aa5-e827363276a8.png)

![download (11)](https://user-images.githubusercontent.com/43179913/125775464-c8fb98ab-b6f2-4380-87ab-627c80eb1c4f.png)

![download (12)](https://user-images.githubusercontent.com/43179913/125775885-7a454ee0-960b-4a83-a92f-f859ae42e04c.png)

**Comment by proofconstruction** [comment]
These look good, especially the last 3. In the last one, I think it would look better if the border effect continued down until it went off the page, because right now it looks a bit like part of the paper is missing, or maybe something is overlapping the paper a bit.

**Comment by jboarman** [comment]
The orientation of the side pages probably could rotate such that they align with the page content.

To achieve this effect, perhaps a white or black border could be added to the page content as each side page is added and extends the canvas of the image to accommodate the additional page line?  In other words, we extend outward instead of creeping inward on the page?

![image](https://user-images.githubusercontent.com/1108065/125833300-7889a354-2504-42b6-8826-c74bafa04c0f.png)

**Comment by kwcckw** [comment]
> 
> 
> Here are some examples with some noise and offset added to the border

I like the effect where you added those noises to the border, it looks more real.

**Comment by shaheryar1** [comment]
> To achieve this effect, perhaps a white or black border could be added to the page content as each side page is added and extends the canvas of the image to accommodate the additional page line? In other words, we extend outward instead of creeping inward on the page?

Great suggestion. Currently I was using the page itself to draw these borders but I guess the appropriate way would be adding an external border. Moreover, let me try the outside extension, it may give us more realistic effects.

**Comment by shaheryar1** [comment]
I have changed it in a way that the content of first page is always aligned with the borders.


![download (7)](https://user-images.githubusercontent.com/43179913/125935069-e96a89c5-e627-4887-850d-86235fc7e650.png)
![download (2)](https://user-images.githubusercontent.com/43179913/125936991-8397aac8-7560-48cb-bf38-edfff7b24733.png)

**Comment by jboarman** [comment]
These look great!

**Comment by shaheryar1** [comment]
I'll go ahead and make a PR for this tomorrow.

**Comment by shaheryar1** [comment]
Here is the notebook where you can see the progress.
https://colab.research.google.com/drive/104dKLjXmG0qg15naxcnF1YgpIHjM9Jh-?usp=sharing

**Comment by jboarman** [comment]
@shaheryar1 Thanks for sharing that NB ... this is definitely the way we should approach testing out new ideas and sharing them with the team.  I ran into an error when I played the notebook using "Run All", but maybe that was just a temporary issue.

**Comment by kwcckw** [comment]
> 
> 
> @shaheryar1 Thanks for sharing that NB ... this is definitely the way we should approach testing out new ideas and sharing them with the team. I ran into an error when I played the notebook using "Run All", but maybe that was just a temporary issue.

There's some error in the code due to `ksize` is not parsed into `add_corner_noise`. I re-edited the code a bit and here's the working version:
https://colab.research.google.com/drive/1PXLmnhc4T6rS7cw6x6bX9Ur62NBdpB9w?usp=sharing

![image](https://user-images.githubusercontent.com/52521979/126053246-d783eaca-0ae8-40c3-bf1d-8ef2b3afec32.png)

**Comment by proofconstruction** [comment]
@shaheryar1 Can you submit a pull request with this augmentation?

**Comment by shaheryar1** [comment]
The pull request #42  has been created for this. I think it might contain some bugs related to image dimensions. Do share your feedback after testing.

**Comment by proofconstruction** [comment]
Merged PR #42
